autosklearn/automl.py

```diff
@@ -429,11 +429,19 @@ def _do_dummy_prediction(self, datamanager: XYDataManager, num_run: int) -> int:
                 )
         return num_run
 
+    @classmethod
+    def _task_type_id(cls, task_type: str) -> int:
+        raise NotImplementedError
+
+    @classmethod
+    def _supports_task_type(cls, task_type: str) -> bool:
+        raise NotImplementedError
+
     def fit(
         self,
         X: SUPPORTED_FEAT_TYPES,
         y: SUPPORTED_TARGET_TYPES,
-        task: int,
+        task: Optional[int] = None,
```

Comment:
Is passing the task type still necessary? It appears that in the calls of the subclass you stopped passing the task type to the super class.

Reply:
The tests in`test/test_automl.py` rely on creating instances of `AutoML`. In this case a `task` must be passed in.

Long story short, I tried to remove it with a workaround for the problem described below but tests started failing and the solution to remove it is a bigger PR, the goal was jsut to fix sparse y.

#### Problem
`AutoML` can't determine the task type for itself. The problem is that `type_of_target(y)` can return `'multiclass'` which we map to either `REGRESSION` or `MULTICLASS_CLASSIFICATION`, depending on the subclass. Hence, when the object is not a subclass, `AutoML` needs this information passed in.

You're right though, it's not really needed, especially not `task` and `is_classification` parameters. The minimal amount of information needed is simply `is_classification: bool`. This also removes the need for subclasses at all.
* `AutoML` can uniquely determine the task using `type_of_target` and `is_classification`
* This would mean the subclasses do nothing except `AutoMLClassifier` which does extra work in `predict` to transform predictions to labels
* This transformation could happen in `AutoSklearnClassifier`, leaving `AutoML` to simply generate outputs, removing the need for having subclasses of `AutoML`

Reply:
Got it, so for now we can leave everything as it is? But could you please add a docstring on why this is necessary?

Reply:
Yes, I say we leave it as is and we can talk about this next week. I've added a docstring to the function with the explanation included.